flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java

```diff
@@ -0,0 +1,270 @@
+				public FnDataReceiver<WindowedValue<byte[]>> create(String pCollectionId) {
+					return input -> {
+						bais.setBuffer(input.getValue(), 0, input.getValue().length);
+						resultReceiver.accept(outputCoder.decode(bais));
+					};
+				}
+			};
+
+		try {
+			remoteBundle = stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler);
+			mainInputReceiver =
+				Preconditions.checkNotNull(
+					remoteBundle.getInputReceivers().get(MAIN_INPUT_ID),
+					"Failed to retrieve main input receiver.");
+		} catch (Throwable t) {
+			throw new RuntimeException("Failed to start remote bundle", t);
+		}
+	}
+
+	@Override
+	public void finishBundle() {
+		try {
+			remoteBundle.close();
+		} catch (Throwable t) {
+			throw new RuntimeException("Failed to close remote bundle", t);
+		} finally {
+			remoteBundle = null;
+		}
+	}
+
+	@Override
+	public void processElement(IN element) {
+		try {
+			baos.reset();
+			inputCoder.encode(element, baos);
+			// TODO: support to use ValueOnlyWindowedValueCoder for better performance.
+			// Currently, FullWindowedValueCoder has to be used in Beam's portability framework.
+			mainInputReceiver.accept(WindowedValue.valueInGlobalWindow(baos.toByteArray()));
+		} catch (Throwable t) {
+			throw new RuntimeException("Failed to process element when sending data to Python SDK harness.", t);
+		}
+	}
+
+	public JobBundleFactory createJobBundleFactory(Struct pipelineOptions) throws Exception {
+		return DefaultJobBundleFactory.create(
+			JobInfo.create(taskName, taskName, createEmptyRetrievalToken(), pipelineOptions));
+	}
+
+	/**
+	 * Creates a specification which specifies the portability Python execution environment.
+	 * It's used by Beam's portability framework to creates the actual Python execution environment.
+	 */
+	protected RunnerApi.Environment createPythonExecutionEnvironment() {
+		if (pythonEnv.getExecType() == PythonEnv.ExecType.PROCESS) {
+			final Map<String, String> env = new HashMap<>(2);
+			env.put("python", pythonEnv.getPythonExec());
+
+			return Environments.createProcessEnvironment(
+				"",
+				"",
+				pythonEnv.getPythonWorkerCmd(),
+				env);
+		} else {
+			throw new UnsupportedOperationException(String.format(
+				"Execution type '%s' is not supported.", pythonEnv.getExecType()));
+		}
+	}
+
+	/**
+	 * Creates a {@link ExecutableStage} which contains the Python user-defined functions to be executed
+	 * and all the other information needed to execute them, such as the execution environment, the input
+	 * and output coder, etc.
+	 */
+	public abstract ExecutableStage createExecutableStage();
+
+	/**
+	 * Returns the coder for input elements.
+	 */
+	public abstract Coder<IN> getInputCoder();
+
+	/**
+	 * Returns the coder for execution results.
+	 */
+	public abstract Coder<OUT> getOutputCoder();
+
+	private String createEmptyRetrievalToken() throws Exception {
+		final File retrievalToken = Files.createTempFile("retrieval_token", "json").toFile();
```

Comment:
If this part is running in TaskManager, the better locations to store temporary files is `getContainingTask().getEnvironment().getTaskManagerInfo().getTmpDirectories()`. And maybe we should delete it after finish bundle? At least we should call `retrievalToken.deleteOnExit()` to ensure that it will be deleted when jvm exits.